File: Official Windows Platform Sample/Windows 8 app samples/[VB]-Windows 8 app samples/VB/Windows 8 app samples/Geolocation sample (Windows 8)/VB/backgroundtask/LocationBackgroundTask.vb
'*********************************************************
'
' Copyright (c) Microsoft. All rights reserved.
' THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
' ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
' IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
' PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
'
'*********************************************************

Imports System.Threading
Imports Windows.ApplicationModel.Background
Imports Windows.Devices.Geolocation
Imports Windows.Storage

' A background task always implements the IBackgroundTask interface
Public NotInheritable Class LocationBackgroundTask
    Implements IBackgroundTask

    Private _cts As CancellationTokenSource = Nothing

    ' The Run method is the entry point of a background task
    Public Async Sub Run(taskInstance As IBackgroundTaskInstance) Implements IBackgroundTask.Run
        ' Get the deferral object from the task instance
        Dim deferral As BackgroundTaskDeferral = taskInstance.GetDeferral()
        Dim settings = ApplicationData.Current.LocalSettings

        ' Associate a cancellation handler with the background task
        AddHandler taskInstance.Canceled, AddressOf OnCanceled

        Try
            ' Get cancellation token
            _cts = New CancellationTokenSource()
            Dim token As CancellationToken = _cts.Token

            ' Get a geolocator object
            Dim geolocator As New Geolocator

            ' Carry out the operation
            Dim pos As Geoposition = Await geolocator.GetGeopositionAsync().AsTask(token)

            settings.Values("Status") = "Time: " + pos.Coordinate.Timestamp.ToString()
            settings.Values("Latitude") = pos.Coordinate.Latitude.ToString()
            settings.Values("Longitude") = pos.Coordinate.Longitude.ToString()
            settings.Values("Accuracy") = pos.Coordinate.Accuracy.ToString()
        Catch generatedExceptionName As UnauthorizedAccessException
            settings.Values("Status") = "Disabled"
            settings.Values("Latitude") = "No data"
            settings.Values("Longitude") = "No data"
            settings.Values("Accuracy") = "No data"
        Catch generatedExceptionName As TaskCanceledException
            ' Safely catch the cancelation
        Finally
            _cts = Nothing
            deferral.Complete()
        End Try
    End Sub

    ' Handles background task cancellation
    Private Sub OnCanceled(sender As IBackgroundTaskInstance, reason As BackgroundTaskCancellationReason)
        ' Cancel the active Geoposition request
        If _cts IsNot Nothing Then
            _cts.Cancel()
            _cts = Nothing
        End If
    End Sub
End Class

File: Official Windows Platform Sample/Windows 8 app samples/[C#]-Windows 8 app samples/C#/Windows 8 app samples/Geolocation sample (Windows 8)/C#/backgroundtask/LocationBackgroundTask.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Windows.ApplicationModel.Background;
using Windows.Devices.Geolocation;
using Windows.Storage;

namespace BackgroundTask
{
    public sealed class LocationBackgroundTask : IBackgroundTask
    {
        CancellationTokenSource m_cts = null;

        async void IBackgroundTask.Run(IBackgroundTaskInstance taskInstance)
        {
            BackgroundTaskDeferral deferral = taskInstance.GetDeferral();

            try
            {
                // Associate a cancellation handler with the background task.
                taskInstance.Canceled += new BackgroundTaskCanceledEventHandler(OnCanceled);

                // Get cancellation token
                if (m_cts == null)
                {
                    m_cts = new CancellationTokenSource();
                }
                CancellationToken token = m_cts.Token;

                // Create geolocator object
                Geolocator geolocator = new Geolocator();

                // Make the request for the current position
                Geoposition pos = await geolocator.GetGeopositionAsync().AsTask(token);

                DateTime currrentTime = DateTime.Now;

                WriteStatusToAppdata("Time: " + currrentTime.ToString());
                WriteGeolocToAppdata(pos);
            }
            catch (UnauthorizedAccessException)
            {
                WriteStatusToAppdata("Disabled");
                WipeGeolocDataFromAppdata();
            }
            finally
            {
                m_cts = null;

                deferral.Complete();
            }
        }

        private void WriteGeolocToAppdata(Geoposition pos)
        {
            var settings = ApplicationData.Current.LocalSettings;
            settings.Values["Latitude"] = pos.Coordinate.Latitude.ToString();
            settings.Values["Longitude"] = pos.Coordinate.Longitude.ToString();
            settings.Values["Accuracy"] = pos.Coordinate.Accuracy.ToString();
        }

        private void WipeGeolocDataFromAppdata()
        {
            var settings = ApplicationData.Current.LocalSettings;
            settings.Values["Latitude"] = "";
            settings.Values["Longitude"] = "";
            settings.Values["Accuracy"] = "";
        }

        private void WriteStatusToAppdata(string status)
        {
            var settings = ApplicationData.Current.LocalSettings;
            settings.Values["Status"] = status;
        }

        private void OnCanceled(IBackgroundTaskInstance sender, BackgroundTaskCancellationReason reason)
        {
            if (m_cts != null)
            {
                m_cts.Cancel();
                m_cts = null;
            }
        }
    }
}
